### PR TITLE
feat(ga): the resource of IP address group support new API

### PR DESCRIPTION
**Pull request by ruwenqiang123**
**What this PR does / why we need it**:

The resource of IP address group support new API to implement new functions.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. support add or remove the cidr block in the IP address group.
2. Support the IP address group associate or disassociate listener.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_ip)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccIpAddressGroup_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccIpAddressGroup_basic -timeout 360m -parallel 4
=== RUN   TestAccIpAddressGroup_basic
=== PAUSE TestAccIpAddressGroup_basic
=== CONT  TestAccIpAddressGroup_basic
--- PASS: TestAccIpAddressGroup_basic (44.01s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        44.076s
```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_ip)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccIpAddressGroup_update_cidr"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccIpAddressGroup_update_cidr -timeout 360m -parallel 4
=== RUN   TestAccIpAddressGroup_update_cidr
=== PAUSE TestAccIpAddressGroup_update_cidr
=== CONT  TestAccIpAddressGroup_update_cidr
--- PASS: TestAccIpAddressGroup_update_cidr (66.52s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        66.562s
```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_ip)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccIpAddressGroup_associateListener"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccIpAddressGroup_associateListener -timeout 360m -parallel 4
=== RUN   TestAccIpAddressGroup_associateListener
=== PAUSE TestAccIpAddressGroup_associateListener
=== CONT  TestAccIpAddressGroup_associateListener
--- PASS: TestAccIpAddressGroup_associateListener (156.08s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        156.130s
```

### CODE REVIEW

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 1]
```diff
@@ -302,6 +310,10 @@ func buildUpdateIpAddressGroupBodyParams(d *schema.ResourceData) map[string]inte
 func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	cfg := meta.(*config.Config)
 	region := cfg.GetRegion(d)
+	client, err := cfg.NewServiceClient("ga", region)
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:313]: Can share one with the client below

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 2]
```diff
@@ -330,12 +342,228 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
 
+func addIps(ctx context.Context, d *schema.ResourceData, meta interface{}, client *golangsdk.ServiceClient,
+	rawParams interface{}) error {
+	addIpsHttpUrl := "v1/ip-groups/{ip_group_id}/add-ips"
+	addIpsPath := client.Endpoint + addIpsHttpUrl
+	addIpsPath = strings.ReplaceAll(addIpsPath, "{ip_group_id}", d.Id())
+
+	addIpsOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:400]: delete

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 3]
```diff
@@ -152,6 +147,10 @@ func buildIpAddressOptionBodyParams(rawParams *schema.Set) []map[string]interfac
 func resourceIpAddressGroupCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	cfg := meta.(*config.Config)
 	region := cfg.GetRegion(d)
+	client, err := cfg.NewServiceClient("ga", region)
+	if err != nil {
+		return diag.Errorf("error creating GA client: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go [hunk 4]
```diff
@@ -108,3 +108,417 @@ resource "huaweicloud_ga_address_group" "test" {
 }
 `, name, description)
 }
+
+func TestAccIpAddressGroup_update_cidr(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go]: done

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go [hunk 5]
```diff
@@ -108,3 +108,417 @@ resource "huaweicloud_ga_address_group" "test" {
...
+
+  ip_addresses {
+    cidr        = "192.168.22.0/24"
+    description = "The 21th CIDR block in the IP address group"
+  }
+}
+`, name)
+}
+
+func TestAccIpAddressGroup_associateListener(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go:125]: the associate listener test contain unbind the listener

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 1 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 2 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

**Comment by ruwenqiang123** [comment]
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_ip)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccIpAddressGroup_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccIpAddressGroup_basic -timeout 360m -parallel 4
=== RUN   TestAccIpAddressGroup_basic
=== PAUSE TestAccIpAddressGroup_basic
=== CONT  TestAccIpAddressGroup_basic
--- PASS: TestAccIpAddressGroup_basic (65.57s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        65.614s
```
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_ip)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccIpAddressGroup_associateListener"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccIpAddressGroup_associateListener -timeout 360m -parallel 4
=== RUN   TestAccIpAddressGroup_associateListener
=== PAUSE TestAccIpAddressGroup_associateListener
=== CONT  TestAccIpAddressGroup_associateListener
--- PASS: TestAccIpAddressGroup_associateListener (159.25s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        159.293s
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 6]
```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		for _, v := range rawArray {
+			raw := v.(map[string]interface{})
+			associateListenerOpt.JSONBody = map[string]interface{}{
+				"listener_id": raw["id"],
+				"type":        raw["type"],
+			}
+			_, err := client.Request("POST", associateListenerPath, &associateListenerOpt)
+			if err != nil {
+				return fmt.Errorf("error associate listener: %s", err)
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:469]:
```suggestion
				return fmt.Errorf("error associate listener for IP address group: %s", err)
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 7]
```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+			}
+
+			err = waitingForStateCompleted(ctx, d, meta, d.Timeout(schema.TimeoutUpdate))
+			if err != nil {
+				return fmt.Errorf("error waiting for association listener (%s) to complete: %s", d.Id(), err)
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:474]:
```suggestion
				return fmt.Errorf("error waiting for IP address group association listener (%s) to complete: %s", d.Id(), err)
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 8]
```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+	addIpsOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		rst := make([]map[string]interface{}, len(rawArray))
+		for i, v := range rawArray {
+			raw := v.(map[string]interface{})
+			rst[i] = map[string]interface{}{
+				"cidr":        utils.ValueIngoreEmpty(raw["cidr"]),
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:395]:
```suggestion
				"cidr":        raw["cidr"],
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 9]
```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+				"description": utils.ValueIngoreEmpty(raw["description"]),
+			}
+		}
+
+		addIpsOpt.JSONBody = utils.RemoveNil(map[string]interface{}{
+			"ip_list": rst,
+		})
+		_, err := client.Request("POST", addIpsPath, &addIpsOpt)
+		if err != nil {
+			return fmt.Errorf("error add IP addresses: %s", err)
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:405]:
```suggestion
			return fmt.Errorf("error add IP addresses for IP address group: %s", err)
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 10]
```diff
@@ -330,9 +334,182 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+		}
+
+		err = waitingForStateCompleted(ctx, d, meta, d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return fmt.Errorf("error waiting for addition the IP addresses (%s) to complete: %s", d.Id(), err)
```
Comment (jinyangyang222) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:410]:
```suggestion
			return fmt.Errorf("error waiting for the completion of adding IP address to the IP address group: %s", err)
```

Comment (jinyangyang222) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:410]: d. id() is the IP address group ID, placing it after IP address is wrong

Comment (jinyangyang222) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:410]: Please check for any errors related to it

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 8 above)
Comment (jinyangyang222) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:395]: this field is requied, don't need ingore empty

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go [hunk 11]
```diff
@@ -95,16 +94,370 @@ func TestAccIpAddressGroup_basic(t *testing.T) {
...
+  }
+}
+`, name)
+}
+
+func testIpAddressGroup_basic_step_2(name string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_ga_address_group" "test" {
+  name        = "%[1]s"
+  description = ""
```
[the remark this answers is not in the document]
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go]: done

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go [hunk 12]
```diff
@@ -95,16 +94,370 @@ func TestAccIpAddressGroup_basic(t *testing.T) {
...
+
+  ip_addresses {
+    cidr        = "192.168.19.0/24"
+    description = "The 19th CIDR block in the IP address group"
+  }
+
+  ip_addresses {
+    cidr        = "192.168.20.0/24"
+    description = "The 20th CIDR block in the IP address group"
+  }
```
[the remark this answers is not in the document]
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 6 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 7 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 8 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 9 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 10 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 13]
```diff
@@ -52,14 +57,13 @@ func ResourceIpAddressGroup() *schema.Resource {
...
 				Elem:     ipAddressGroupSchema(),
-				MaxItems: 20,
 			},
-			"id": {
-				Type:     schema.TypeString,
-				Computed: true,
+			"listeners": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MinItems: 1,
```
Comment (ShiChangkuo) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: the MinItems equals required, please delete it.

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 14]
```diff
@@ -330,9 +334,184 @@ func resourceIpAddressGroupUpdate(ctx context.Context, d *schema.ResourceData, m
...
+
+		if removeIpsRaw.Len() > 0 {
+			err = removeIps(ctx, d, meta, updateIpAddressGroupClient, removeIpsRaw.List())
+			if err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		if addIpsRaw.Len() > 0 {
+			err = addIps(ctx, d, meta, updateIpAddressGroupClient, addIpsRaw.List())
```
Comment (ShiChangkuo) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:339]: what will be happened if more than 20 CIDR blocks added at a time?

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 13 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 14 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:339]: done

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go [hunk 15]
```diff
@@ -95,16 +93,289 @@ func TestAccIpAddressGroup_basic(t *testing.T) {
...
+}
+`, name)
+}
+
+func testIpAddressGroup_basic_step_2(name string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_ga_address_group" "test" {
+  name = "%[1]s"
+
+  ip_addresses {
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go]: suggest to use `dynamic` block, for example
```hcl
  dynamic "ip_addresses" {
    for_each = range(0, 10)
    content {
      cidr        = "192.168.${ip_addresses.value}.0/24"
      description = "The ${ip_addresses.value}th IP addresses included in the address group"
    }
  }
```

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 16]
```diff
@@ -52,14 +57,12 @@ func ResourceIpAddressGroup() *schema.Resource {
 			"ip_addresses": {
 				Type:     schema.TypeSet,
 				Optional: true,
-				Computed: true,
-				ForceNew: true,
 				Elem:     ipAddressGroupSchema(),
-				MaxItems: 20,
```
Comment (ShiChangkuo) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:58]: we must process the length of ip_addresses > 20 when creating.

File: huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go (hunk 15 above)
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/resource_huaweicloud_ga_address_group_test.go]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go (hunk 16 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:58]: done

File: huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go [hunk 17]
```diff
@@ -188,11 +165,25 @@ func resourceIpAddressGroupCreate(ctx context.Context, d *schema.ResourceData, m
 
 	d.SetId(id.(string))
 
-	err = createIpAddressGroupWaitingForStateCompleted(ctx, d, meta, d.Timeout(schema.TimeoutCreate))
+	err = waitingForStateCompleted(ctx, d, meta, d.Timeout(schema.TimeoutCreate))
 	if err != nil {
 		return diag.Errorf("error waiting for the creation of IP address group (%s) to complete: %s", d.Id(), err)
 	}
 
+	if val, ok := d.GetOk("ip_addresses"); ok {
```
Comment (ShiChangkuo) [inline huaweicloud/services/ga/resource_huaweicloud_ga_address_group.go:174]:
```suggestion
	//  call addIps to support more than 20 ip addresses
	if val, ok := d.GetOk("ip_addresses"); ok {
```

**Pull request merged**